lib/preflight/preflighttest.js

```diff
@@ -0,0 +1,483 @@
+    let tracksToWaitFor = n - remoteParticipant.tracks.size;
+    if (tracksToWaitFor > 0) {
+      remoteParticipant.on('trackSubscribed', () => {
+        tracksToWaitFor--;
+        if (tracksToWaitFor === 0) {
+          resolve();
+        }
+      });
+    } else {
+      resolve();
+    }
+  });
+}
+
+function runPreflightTest(publisherToken, subscriberToken, options, preflightTest) {
+  const testDuration = options.duration || DEFAULT_TEST_DURATION;
+  delete options.duration; // duration is not a Video.connect option.
+
+  options = Object.assign(options, {
+    video: false,
+    audio: false,
+    networkQuality: true
+  });
+
+  const testTiming = new TimeMeasurement();
+  let localTracks = null;
+  let publisherRoom = null;
+  let subscriberRoom = null;
+  let trackStartListener = null;
+  let connectTiming = null;
+  let mediaTiming = null;
+
+  function collectIceCandidates() {
+    return Promise.resolve().then(() => {
+      const pc = subscriberRoom._signaling._peerConnectionManager._peerConnections.values().next().value._peerConnection._peerConnection;
+      return pc.getStats().then(stats => {
+        return [...stats.values()].filter(stat => {
+          return stat.type === 'local-candidate' || stat.type === 'remote-candidate';
+        });
+      });
+    }).catch(() => {
+      return [];
+    });
+  }
+
+  function collectRTCStats(collectedStats) {
+    return Promise.all([subscriberRoom, publisherRoom].map(room => room._signaling.getStats()))
+      // eslint-disable-next-line consistent-return
+      .then(([subscriberStats, publisherStats]) => {
+        const subscriberStatValues = [...subscriberStats.values()];
+        const publisherStatValues = [...publisherStats.values()];
+
+        if (publisherStatValues.length > 0) {
+          const { activeIceCandidatePair } = publisherStatValues[0];
+          if (activeIceCandidatePair && typeof activeIceCandidatePair.availableOutgoingBitrate === 'number') {
+            collectedStats.outgoingBitrate.push(activeIceCandidatePair.availableOutgoingBitrate);
+          }
+        }
+        if (subscriberStatValues.length > 0) {
+          const { activeIceCandidatePair, remoteAudioTrackStats, remoteVideoTrackStats } = subscriberStatValues[0];
+          if (activeIceCandidatePair) {
+            if (typeof activeIceCandidatePair.totalRoundTripTime === 'number') {
+              collectedStats.rtt.push(activeIceCandidatePair.totalRoundTripTime * 1000);
+            }
+            if (typeof activeIceCandidatePair.availableIncomingBitrate === 'number') {
+              collectedStats.incomingBitrate.push(activeIceCandidatePair.availableIncomingBitrate);
+            }
+
+            if (!collectedStats.selectedIceCandidatePairStats) {
+              collectedStats.selectedIceCandidatePairStats = {
+                localCandidate: activeIceCandidatePair.localCandidate,
+                remoteCandidate: activeIceCandidatePair.remoteCandidate
```

Comment:
`activeIceCandidatePair.remoteCandidate` and `activeIceCandidatePair.localCandidate` is of type [RTCIceCandidateStats](https://developer.mozilla.org/en-US/docs/Web/API/RTCIceCandidateStats) and not [RTCIceCandidate](https://developer.mozilla.org/en-US/docs/Web/API/RTCIceCandidate) correct?

Reply:
yes they are of type `RTCIceCandidateStats`